spec/uvarint_spec.cr

```diff
@@ -1,77 +1,191 @@
-    buf = [0x80_u8, n]
-    d = UVarint.decode buf
-    d.should eq(n.to_u64 << 7)
+  describe "decode" do
+    it "decodes single bytes" do
+      n = Random.rand(0..0x7F).to_u8
+      b = Bytes[n]
+      v = UVarInt.new b
+      u = v.uint
+      u.should eq(n)
+    end
+
+    it "decodes multiple bytes" do
+      b = Bytes[0xAC_u8, 0x02_u8]
+      v = UVarInt.new b
+      u = v.uint
+      u.should eq(300)
+    end
+
+    it "decodes multiple bytes with zero" do
+      n = Random.rand(0x7F).to_u8
+      b = Bytes[0x80_u8, n]
+      v = UVarInt.new b
+      u = v.uint
+      u.should eq(n.to_u64 << 7)
+    end
   end
 
-  it "encodes to single bytes" do
-    n = Random.rand(0x7F).to_u64
-    e = UVarint.encode n
-    e.should eq([n.to_u8])
+  describe "encode" do
+    it "encodes to single bytes" do
+      n = Random.rand(0x7F).to_u64
+      v = UVarInt.new n
+      b = v.bytes
+      b.should eq(Bytes[n.to_u8])
+    end
+
+    it "encodes to multiple bytes" do
+      v = UVarInt.new 300_u64
+      b = v.bytes
+      b.should eq(Bytes[0xAC_u8, 0x02_u8])
+    end
+
+    it "encodes to multiple bytes with first byte zero" do
+      v = UVarInt.new 0x0F00_u64
+      b = v.bytes
+      b.should eq(Bytes[0x80_u8, 0x1E_u8])
+    end
   end
 
-  it "encodes to multiple bytes" do
-    e = UVarint.encode 300_u64
-    e.should eq([0xAC_u8, 0x02_u8])
+  describe "to_s" do
+    it "returns self.uint.to_s" do
+      v = UVarInt.new 300_u64
+      s = v.to_s 16
+      s.should eq("12c")
+    end
   end
 
-  it "encodes to multiple bytes with first byte zero" do
-    n = 0x0F00_u64
-    e = UVarint.encode n
-    e.should eq([0x80_u8, 0x1E_u8])
+  describe "hexstring" do
+    it "returns self.bytes.hexstring" do
+      v = UVarInt.new 300_u64
+      h = v.hexstring
+      h.should eq("ac02")
+    end
   end
 
-  it "encodes and decodes big integers" do
-    bigs = [] of UInt64
-    (32..53).each do |i|
-      n = (2 ** i).to_u64
-      bigs << n - 1
-      bigs << n
+  describe "fuzzing" do
+    it "encodes and decodes random values" do
+      (0..100000).each do
+        n = Random.rand(0..0x7FFFFFFF).to_u64
+        v = UVarInt.new n
+        u = v.uint
+        u.should eq(n)
+      end
     end
 
-    bigs.each do |n|
-      e = UVarint.encode n
-      d = UVarint.decode e
-      d.should eq(n)
+    it "encodes and decodes big integers" do
+      bigs = [] of UInt64
+      (32..53).each do |i|
+        n = (2 ** i).to_u64
+        bigs << n - 1
+        bigs << n
+      end
+
+      bigs.each do |n|
+        v = UVarInt.new n
+        u = v.uint
+        u.should eq(n)
+      end
     end
-  end
 
-  it "encodes and decodes really big integers" do
-    max_u32 = 0_u32..0xFFFFFFFF_u32
-    (0..100000).each do
-      upper = Random.rand(max_u32).to_u64 << 32
-      lower = Random.rand(max_u32).to_u64
-      n = upper + lower
-      e = UVarint.encode n
-      d = UVarint.decode e
-      d.should eq(n)
+    it "encodes and decodes really big integers" do
```

Comment:
"Really big" just means a random `UInt64` between the max value of `UInt32` and `UInt64`. `Random.rand` only accepts up to an `Int32` for the range, so I had to generate the lower and upper 32 bits individually. I will clarify this in the description and perhaps with a comment.

If we add support for `BigInt`, we could add a test for that here too.